### DISCUSSION TITLE
What happened to development?

### DISCUSSION THREAD

**Discussion by kopeboy** [dogecoin/dogecoin#3392] [General]
Why hasn't there been any commit since 1.5 years? 
Is the development of Dogecoin happening elsewhere, in private, now or has it stopped?
Thank you.

**Comment by patricklodder** [comment]
Development is happening on `-dev` branches, see https://github.com/dogecoin/dogecoin#branches for rationale. Only when we do a release, the `master` branch will be updated.

For the next minor release, see [1.14.7-dev](https://github.com/dogecoin/dogecoin/tree/1.14.7-dev).

**Reply by kopeboy** [reply]
Thank you and sorry for the noise

**Reply by patricklodder** [reply]
No probs! It's good to have this question documented here.

**Reply by asidesss** [reply]
U tell me

**Comment by patiee** [comment]
When do you plan to add smart contracts to dogecoin? 
Which language do you want to use for smart contracts - solidity or cosmwasm?

**Reply by patricklodder** [reply]
I'll help anyone trying to do this on L2 or sidechain/x-chain (so that you can do solidity or cosmwasm or any other VM) under 2 conditions:

1. Must have a clear demonstration of how it will operate securely in a scenario without scams
2. Should not introduce another gas token / ICO / airdrop / other scammy solutions

There was a sidechain based on polygon in the past that was cool, except they messed up the bridge and proceeded to launch a scam token as a gas token. This means that honest, skilled actors can actually do it today. I'm still waiting for non-scammers to step up.

**Reply by patiee** [reply]
I have a lot of experience with indexing blockchains written in Go, I can try to help with integration, but I need to take a look on dogecoin implementation first. Final solution shouldn't have gas token for transaction as the main requirement?

Smart contact code is executed on validator node - there's need to migrate from POW to POS in the near future.

**Reply by patricklodder** [reply]
> Final solution shouldn't have gas token for transaction as the main requirement?

Exactly. This is because introducing a new gas token makes the entire solution speculative. I'd also like to see the business case of how, when you take out all the scam use-cases, there can be an expectation for fees to be sustainable, or for fees to not be needed because operators of the envisioned L2/sidechain have another reason to operate a honest node on that network.

> there's need to migrate from POW to POS in the near future.

The first EVM chain didn't have PoS either, and ETC still is POW too... so this is a provable misconception. Additionally, you do not want to change L1 consensus to introduce this functionality. Bridging can be done with for example DSS, like Thorchain uses for vaults.

**Reply by patiee** [reply]
Yes, POS is not required for smart contracts. I need to study dogecoin code first more to tell how this could be implemented in details. Without gas token blockchain needs to have different system for smart contract fee. Maybe for now this could be just constant variables for:
- store - push code to dogecoin network
- instantiate - initialize smart contract with configuration
- execute - make changes inside smart contract 
- query - this could be free

**Reply by patiee** [reply]
I have a lot of experience with indexing blockchain. My main goal will be to implement the fastest storage for transaction queries that ever existed in C++. It doesn't make much sense at this moment because there's no smart contracts. 

Const variables for smart contracts fees could be added to genesis. I was never using C++ before in blockchain, this week I just study the code to see if it will be better to implement smart contract integration from scratch rather then using Solidity or CosmWasm 👀 

Solidity is written in C++ https://github.com/ethereum/solidity

**Reply by patiee** [reply]
Personally I don't like internal transactions from evm

**Reply by patiee** [reply]
Current block structure is just with simple transaction data. Smart contracts are emitting events - cosmos-sdk have events in 3 different places inside Block. Evm emits events in internal transactions.

**Reply by patiee** [reply]
cosmos-sdk emits events already decoded, with evm there's need to load smart contract registry first to decode content - which approach do you like more?

**Reply by patiee** [reply]
My preference is CosmWasm - I like decoded events data inside Block structure and this is very simple to deploy smart contract. Rust is maybe not the easiest language to write smart contracts, but to deploy erc-20 there's many examples that you can clone - [cw-tokens](https://github.com/CosmWasm/cw-tokens). I'll try to make a new transaction type to store wasm code.

**Reply by patiee** [reply]
I cloned repository today and added smart contract fees to consensus 🎉

https://github.com/dogecoin/dogecoin/compare/master...patiee:dogecoin:smart-contracts

Now I'm trying to find the best way to create a new transaction type to store wasm code

**Reply by patricklodder** [reply]
Why would you do that on L1 though?
